client/state/index.js

```diff
@@ -111,26 +115,113 @@ export const reducer = combineReducers( {
 	ui,
 	users,
 	wordads,
-} );
-
-export function createReduxStore( initialState = {} ) {
-	const isBrowser = typeof window === 'object';
-	const isAudioSupported = typeof window === 'object' && typeof window.Audio === 'function';
-
-	const middlewares = [
-		thunkMiddleware,
-		noticesMiddleware,
-		isBrowser && require( './analytics/middleware.js' ).analyticsMiddleware,
-		isBrowser && require( './data-layer/wpcom-api-middleware.js' ).default,
-		isAudioSupported && require( './audio/middleware.js' ).default,
-	].filter( Boolean );
-
-	const enhancers = [
-		isBrowser && window.app && window.app.isDebug && consoleDispatcher,
-		applyMiddleware( ...middlewares ),
-		isBrowser && sitesSync,
-		isBrowser && window.devToolsExtension && window.devToolsExtension()
-	].filter( Boolean );
-
-	return compose( ...enhancers )( createStore )( reducer, initialState );
+};
+
+/**
+ * Creates and holds the Redux Store and its dependencies.
+ * @param { object } reducers: The named set of reducers for the store.
+ */
+export class CalypsoStore {
+	constructor() {
+		this.reducers = { ...initialReducers };
+		this.reduxStore = null;
+	}
+
+	createReduxStore( initialState = {} ) {
+		const isBrowser = typeof window === 'object';
+		const isAudioSupported = typeof window === 'object' && typeof window.Audio === 'function';
+
+		const middlewares = [
+			thunkMiddleware,
+			noticesMiddleware,
+			isBrowser && require( './analytics/middleware.js' ).analyticsMiddleware,
+			isBrowser && require( './data-layer/wpcom-api-middleware.js' ).default,
+			isAudioSupported && require( './audio/middleware.js' ).default,
+		].filter( Boolean );
+
+		const enhancers = [
+			isBrowser && window.app && window.app.isDebug && consoleDispatcher,
+			applyMiddleware( ...middlewares ),
+			isBrowser && sitesSync,
+			isBrowser && window.devToolsExtension && window.devToolsExtension()
+		].filter( Boolean );
+
+		const reducer = this.getCombinedReducer();
+		this.reduxStore = compose( ...enhancers )( createStore )( reducer, initialState );
+		return this.reduxStore;
+	}
+
+	addReducer( name, reducerFunc ) {
+		if ( this.reducers[ name ] !== undefined ) {
+			throw new Error( `addReducer(): name ${ name } already in use` );
+		}
+
+		this.setReducers( { ...this.reducers, [ name ]: reducerFunc } );
+	}
+
+	removeReducer( name ) {
+		const { [ name ]: removedReducer, ...remainingReducers } = this.reducers;
+
+		if ( ! removedReducer ) {
+			debug( 'removeReducer(): name not found' );
+			return;
+		}
+
+		this.setReducers( remainingReducers );
+	}
+
+	getCombinedReducer() {
+		const reducers = this.reducers;
+		return combineReducers( reducers );
+	}
```

Comment:
Can we memoize this function based on the reference of `this.reducers`? Since it gets called in a number of different places, repeatedly calling `combineReducers` seems avoidable.